azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -79,18 +86,39 @@ def decrypt_p7m(self, p7m_file, trans_prv_file, trans_cert_file, pem_file):
         elif not os.path.exists(trans_prv_file):
             raise IOError(errno.ENOENT, "File not found", trans_prv_file)
         else:
-            cmd = ("{0} cms -decrypt -in {1} -inkey {2} -recip {3} "
-                   "| {4} pkcs12 -nodes -password pass: -out {5}"
-                   "").format(self.openssl_cmd, p7m_file, trans_prv_file,
-                              trans_cert_file, self.openssl_cmd, pem_file)
-            shellutil.run(cmd)
-            rc = shellutil.run(cmd) # pylint: disable=C0103
-            if rc != 0:
-                logger.error("Failed to decrypt {0}".format(p7m_file))
+            first_cmd = [self.openssl_cmd, "cms", "-decrypt", "-in", p7m_file, "-inkey",
+                trans_prv_file, "-recip", trans_cert_file]
+            second_cmd = [self.openssl_cmd, "pkcs12", "-nodes", "-password", "pass:", 
+                "-out", pem_file]
+
+            first_proc = subprocess.Popen(first_cmd, stdout=subprocess.PIPE)
+
+            second_proc = subprocess.Popen(second_cmd, stdin=first_proc.stdout, stdout=subprocess.PIPE)
+            first_proc.stdout.close()  # see https://docs.python.org/2/library/subprocess.html#replacing-shell-pipeline
+            stdout, stderr = second_proc.communicate()
+            first_proc.poll()  # ensure that returncode is initialized
+
+            if first_proc.returncode != 0 or second_proc.returncode != 0:
+                stdout = ustr(stdout, encoding='utf-8', errors="backslashreplace") if stdout else ""
+                stderr =  ustr(stderr, encoding='utf-8', errors="backslashreplace") if stderr else ""
+
+                msg = """Failed to decrypt {0}
+                [stdout]
+                {1}
+
+                [stderr]
+                {2}
+                """.format(p7m_file, stdout, stderr)
+                logger.error(msg)
+            
 
     def crt_to_ssh(self, input_file, output_file):
-        shellutil.run("ssh-keygen -i -m PKCS8 -f {0} >> {1}".format(input_file,
-                                                                    output_file))
+        with open(output_file, "ab") as file_out:
+            cmd = ["ssh-keygen", "-i", "-m", "PKCS8", "-f", input_file]
+            keygen_proc = subprocess.Popen(cmd, stdout=file_out)
+            keygen_proc.wait()
```

Comment:
This is a bare-bones fix to prove the DCR issue we're running in to.

Reply:
This issue here was that `subprocess.communicate` was returning `None` for `stdout` in a prior revision; a subsequent `file_out.write` call would therefore fail, crashing the provisioning part of the agent.